File: 8-string-to-integer-(atoi).py
class Solution:
    def myAtoi(self, s: str) -> int:
        s = s.strip()

        if not s:
            return 0

        i = 0
        sign = 1

        if s[i] == "-":
            sign = -1
            i += 1
        elif s[i] == "+":
            i += 1

        ans = 0

        while i < len(s) and s[i].isdigit():
            val = ans * 10 + int(s[i])
            if val >= 2**31:
                ret = (2**31) * sign
                return ret - 1 if sign > 0 else ret

            ans = val
            i += 1

        return ans * sign

File: 8-string-to-integer-(atoi).java
class Solution {
    public int myAtoi(String s) {
        s = s.trim();
        if (s.isEmpty())
            return 0;

        int sign = 1, i = 0;

        if (s.charAt(i) == '-') {
            sign = -1;
            i++;
        } else if (s.charAt(i) == '+')
            i++;

        long ans = 0;

        while (i < s.length() && Character.isDigit(s.charAt(i))) {
            long val = ans * 10 + (s.charAt(i) - '0');
            if (val > 2147483647) {
                return sign > 0 ? 2147483647 : -2147483648;
            }
            ans = val;
            i++;
        }
        return (int) (ans * sign);
    }
}
